src/include/distributed/multi_physical_planner.h

```diff
@@ -182,15 +182,23 @@ typedef struct Task
 	uint32 taskId;
 
 	/*
-	 * If queryString != NULL, then we have a single query for all placements.
-	 * Otherwise, length of perPlacementQueryStrings is equal to length of
-	 * taskPlacementList and can assign a different query for each placement.
-	 * We need this flexibility when a query should return node specific values.
-	 * For example, on which node did we succeed storing some result files?
+	 * For most queries either query or queryStringLazy is not NULL. This means
+	 * we have a single query for all placements.
+	 * If this is not the case, the length of perPlacementQueryStrings is
+	 * non-zero and equal to length of taskPlacementList. Like this it can
+	 * assign a different query for each placement. We need this flexibility
+	 * when a query should return node specific values. For example, on which
+	 * node did we succeed storing some result files?
+	 *
+	 * In almost all cases queryStringLazy should be read by using
+	 * TaskQueryString(). This will populate the field if only the query field
+	 * is not NULL.
 	 */
-	char *queryString;
+	Query *query;
+	char *queryStringLazy;
 	List *perPlacementQueryStrings;
 
+	Oid distributedTableId;     /* only applies to insert tasks that are locally executed */
```

Comment:
I can see that this'd be useful and many devs will use this. How hard is it to make this generic for all tasks?

Reply:
removed it and replaced it with relationShardList, like discussed in person

Reply:
Turns out my solution with relationShardList was not working for `INSERT ... SELECT` queries. I'v changed it back to distributedTableId now, but renamed `distributedTableId` to `anchorDistributedTableId`. I think this is the easiest solution.